Fix FOR bug, also fix bench to compile

a10y opened the pull request:
@robert3005 

1. Fix issue with FoR compression where nullability was getting stripped, causing `compress` benchmark to fail
2. Fix some random tokio bs that was also causing `compress` benchmark to fail due to recursive runtime creations

a10y committed:
fix compress benchmark

a10y commented on vortex-fastlanes/src/for/compress.rs:
```diff
@@ -53,7 +54,7 @@ impl EncodingCompression for FoREncoding {
         let child = match_each_integer_ptype!(parray.ptype(), |$T| {
             if shift == <$T>::PTYPE.bit_width() as u8 {
-                ConstantArray::new($T::default(), parray.len()).into_array()
+                ConstantArray::new(Scalar::zero::<$T>(parray.dtype().nullability()), parray.len()).into_array()
```
propagate nullability from child

a10y commented on bench-vortex/src/reader.rs:
```diff
@@ -47,6 +47,17 @@ pub fn open_vortex(path: &Path) -> VortexResult<Array> {
         .map(|a| a.into_array())
 }
 
+pub async fn open_vortex_async(path: &Path) -> VortexResult<Array> {
```
this is just copy-paste but without the block_on

robert3005 replied:
I really wish there was a way not to fork these functions... but I think this is fixed once we have some higher level abstractions. Or benchmarks are just tokio.

a10y committed:
save

a10y committed:
lord give me the strength

a10y committed:
change back

a10y commented on bench-vortex/benches/random_access.rs:
```diff
@@ -27,6 +27,7 @@ fn random_access(c: &mut Criterion) {
     });
 
     let dataset = BenchmarkDatasets::PBI(Medicare1);
+    dataset.write_as_parquet();
```
this needs to happen to convert the CSVs to parquet, so that the next line can convert the parquet's to lance

a10y commented on vortex-alp/src/compute.rs:
```diff
@@ -61,3 +68,52 @@ impl SliceFn for ALPArray {
+
+        let flat = alp.as_contiguous(&[encoded]).unwrap();
+
+        let a = scalar_at(&flat, 0)
+            .unwrap()
+            .value()
+            .as_pvalue()
+            .unwrap()
+            .unwrap();
+        let a: f64 = a.try_into().unwrap();
+
+        let b = scalar_at(&flat, 1)
+            .unwrap()
+            .value()
+            .as_pvalue()
+            .unwrap()
+            .unwrap();
+        let b: f64 = b.try_into().unwrap();
+
+        let c = scalar_at(&flat, 2)
+            .unwrap()
+            .value()
+            .as_pvalue()
+            .unwrap()
+            .unwrap();
+        let c: f64 = c.try_into().unwrap();
+
+        assert_eq!(a, 1.0);
+        assert_eq!(b, 2.0);
+        assert_eq!(c, 3.0);
```
I'm not sure if there's a more concise way to get the `f64`s back out of the flattened thing

gatesn replied:
Can have the test `-> VortexResult<()>` and then use `?`. I think you can also just try_into on the scalar without unwrapping as a pvalue

a10y replied:
i ended up just going with the try_into() that rob pointed out below

a10y commented on vortex-array/src/array/datetime/localdatetime.rs:
```diff
@@ -54,6 +54,26 @@ impl LocalDateTimeArray {
     }
 }
 
+impl TryFrom<LocalDateTimeArray> for ExtensionArray {
```
this was missing to get from the LocalDateTimeArray back to the ExtensionArray, so we can return a LocalDateTimeArray as a flattened thing

a10y commented on vortex-array/src/compress.rs:
```diff
@@ -226,6 +215,39 @@ impl<'a> Compressor<'a> {
 
+/// Check that compression did not alter the length of the validity array.
+fn check_validity_unchanged(arr: &Array, compressed: &Array) {
+    let _ = arr;
+    let _ = compressed;
```
needed to prevent compiler complaining about unused vars during release compiles

a10y commented on vortex-datetime-parts/src/compute.rs:
```diff
@@ -39,3 +54,133 @@ impl SliceFn for DateTimePartsArray {
+        .iter()
+        .zip(seconds_buf.iter())
+        .zip(subsecond_buf.iter())
+        .map(|((d, s), ss)| d * 86_400 * divisor + s * divisor + ss)
```
this logic is wrong, lemme fix it real quick

a10y replied:
nvm it's correct. added a test though

a10y committed:
cleanup ,test

a10y commented on vortex-fastlanes/src/for/compute.rs:
```diff
@@ -86,4 +94,31 @@ mod test {
     }
+
+    #[test]
+    fn for_as_contiguous() {
```
i guess we should probably centralize 1-2 smoketests for each of the `ArrayCompute` functions for all of the variants.

a10y commented on vortex-datetime-parts/src/compute.rs:
```diff
@@ -39,3 +54,201 @@ impl SliceFn for DateTimePartsArray {
+
+        let local = decode_to_localdatetime(&date_times.into_array()).unwrap();
+
+        let elem0: i64 = scalar_at(&local.timestamps(), 0)
```
this is gross and if there is a cleaner way i will happily do it

a10y replied:
fixed

a10y committed:
whoops

a10y commented:
Ok, I think I fixed all of the errors in the benchmarks now.

1. Implemented `AsContiguousFn` for all of the types that get touched by the taxi dataset. I added an `impl_default_as_contiguous!` macro so that types can easily opt themselves in to implementing AsContiguous in terms of flattens. I wanted to do a blanket implementation but you can't override blanket implementations :(
1. Fixed some bugs in how StructArray as_contiguous was implemented
1. Fixed some bugs in how
1. Fix how medicare data was being written

Now the benchmarks complete without error which is nice. I can FLUP with a github action to run the benchmarks in a scheduled action to get faster feedback if things break

a10y committed:
sum

a10y committed:
fix clippy

a10y committed:
more

a10y committed:
add impl TryFrom<Scalar> for $T

robert3005 commented on bench-vortex/src/public_bi_data.rs:
```diff
@@ -503,6 +503,7 @@ impl BenchmarkDataset for BenchmarkDatasets {
         self.as_uncompressed();
         for f in self.list_files(FileType::Csv) {
             info!("Compressing {} to lance", f.to_str().unwrap());
+            println!("Compressing {} to lance", f.to_str().unwrap());
```
kill this line?

robert3005 commented on vortex-array/src/compute/as_contiguous.rs:
```diff
@@ -3,18 +3,49 @@ use vortex_error::{vortex_bail, vortex_err, VortexResult};
+                    vortex_error::vortex_bail!(ComputeError: "mismatched dtypes in call to as_contiguous");
+                }
+
+                let mut chunks = Vec::with_capacity(arrays.iter().map(|array| array.len()).sum());
```
this is too big - you want arrays.len() + 1 and not sum of all array lengths

robert3005 commented on vortex-alp/src/compute.rs:
```diff
@@ -61,3 +68,52 @@ impl SliceFn for ALPArray {
+
+        let flat = alp.as_contiguous(&[encoded]).unwrap();
+
+        let a = scalar_at(&flat, 0)
```
There's `TryFrom<Scalar> for f64` same for &Scalar so you can just do `scalar_at(&flat, 0).unwrap().try_into().unwrap()` here

robert3005 commented on vortex-array/src/array/struct/compute.rs:
```diff
@@ -92,9 +96,19 @@ impl AsContiguousFn for StructArray {
+                        .cloned()
+                        .map(|array| array.flatten().unwrap().into_array())
+                        .collect::<Vec<_>>();
+                    as_contiguous(flattened.as_slice())
```
This is the wrong place to do the flattening imho. Here we should just delegate

robert3005 replied:
i.e. we should just call `as_contiguous`

a10y replied:
yea it's just that as_contiguous throws up if you pass it arrays with different encodings.

are you suggesting to fix as_contiguous to handle that case better?

robert3005 replied:
Oh… sorry seems I have gone in circles. This is just broken api. Leave this as is for now

robert3005 commented on vortex-datetime-parts/src/compute.rs:
```diff
@@ -39,3 +54,193 @@ impl SliceFn for DateTimePartsArray {
+            vortex_bail!(ComputeError: "mismatched dtypes in call to as_contiguous");
+        }
+
+        let mut chunks = Vec::with_capacity(arrays.iter().map(|array| array.len()).sum());
```
Same issue as struct array

robert3005 commented on vortex-datetime-parts/src/compute.rs:
```diff
@@ -39,3 +54,193 @@ impl SliceFn for DateTimePartsArray {
+                    TimeUnit::S => 1,
+                };
+
+                let days = scalar_at(&self.days(), index)?.into_value();
```
I think you will always get i64 here? If not we can switch over ptype and unwrap directly into the necessary type

a10y committed:
comments

a10y committed:
simplify unwrapping scalar

a10y committed:
clippy

gatesn commented:
Discussed offline about removing as_contiguous and reaffirming how flatten should work

a10y committed:
fix array len

a10y commented:
I believe this PR now addresses all of the stylistic comments.

For the AsContiguous removal: I broke that out into a separate PR stacked on this branch: #346 

Would love a review there of the structure, and maybe a discussion about how best to handle List and Extension dtyped-arrays

a10y committed:
Remove AsContiguousFn (#346)

Per the comments on #341 , this PR removes the AsContiguousFn trait and
as_contiguous function entirely, and instead pushes the relevant details
into the implementation of `ArrayFlatten::flatten` for the
`ChunkedArray` type.

A few questions I've bumped into while doing this refactor:

* Can DType::List only reference primitive types? I see ListScalar, but
it’s not clear to me how Vortex would encode e.g. `List<Struct>` or
`List<List>`
* **Answer**: Deferring for now as List DTypes not fully support yet
anyway
* How should this work for ExtensionArray?
* **Answer**: ExtensionArray can contain a ChunkedArray for its internal
storage Array

a10y committed:
Merge remote-tracking branch 'origin/develop' into aduffy/fix-bench
